File: 05 May/05-10/isbn_13Validator.js
function isValidIsbn13(str) {
  if (/[^\d-]/.exec(str)) {
    return false;
  }

  const strippedOfDashes = str.replaceAll("-", "");

  if (strippedOfDashes.length !== 13) {
    return false;
  }

  return !(strippedOfDashes
    .split("")
    .reduce((acc, digit, index) => acc + digit * (index % 2 ? 3 : 1), 0) 
    % 10
  );
}

// --- TEST SUITE ---

const testsText = String.raw`
1. isValidIsbn13("9780306406157") should return true.
2. isValidIsbn13("97803064061570") should return false.
3. isValidIsbn13("978-0-13-595705-9") should return true.
4. isValidIsbn13("978-030-64061A-4") should return false.
5. isValidIsbn13("9-7-8-0-1-3-4-7-5-7-5-9-9") should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 05 May/05-10/isbn_13_validator.py
def is_valid_isbn_13(s):
    if re.search("[^\d-]", s):
        return False

    stripped_of_dashes = s.replace("-", "")

    if len(stripped_of_dashes) != 13:
        return False

    return not sum([int(digit) * (3 if index % 2 else 1) for index, digit in enumerate(list(stripped_of_dashes))]) % 10

# --- TEST SUITE ---

tests_text = r'''
1. is_valid_isbn_13("9780306406157") should return True.
2. is_valid_isbn_13("97803064061570") should return False.
3. is_valid_isbn_13("978-0-13-595705-9") should return True.
4. is_valid_isbn_13("978-030-64061A-4") should return False.
5. is_valid_isbn_13("9-7-8-0-1-3-4-7-5-7-5-9-9") should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)